File: spikes/02-binding-toolchain/harness/bench.mjs
/**
 * The one JavaScript timing harness of spike 2, shared by the option A and
 * option B benchmarks so their rows are measured the same way.
 *
 * A row is the median of `iterations` timed calls after a warm-up, in
 * microseconds, with the 90th percentile beside it.
 */

import { mkdirSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';

/**
 * Times `fn` over `rounds` independent rounds and keeps the round with the
 * lowest median, so a garbage-collection pause or a late optimisation in
 * one round does not stand for the cost of the call.
 */
export function bench(name, fn, { iterations = 2000, warmup = 200, rounds = 3 } = {}) {
  let best = null;
  for (let round = 0; round < rounds; round += 1) {
    for (let i = 0; i < warmup; i += 1) fn();
    const times = new Float64Array(iterations);
    for (let i = 0; i < iterations; i += 1) {
      const start = performance.now();
      fn();
      times[i] = performance.now() - start;
    }
    times.sort();
    const at = (q) => times[Math.min(iterations - 1, Math.floor(q * iterations))] * 1000;
    const row = { name, median_us: at(0.5), p90_us: at(0.9), iterations, rounds };
    if (best === null || row.median_us < best.median_us) best = row;
  }
  return best;
}

/** A row the binding cannot express, with the reason. */
export function unavailable(name, note) {
  return { name, median_us: null, p90_us: null, iterations: 0, note };
}

/** A row derived from two others: `(whole − base) / divisor`. */
export function derived(name, whole, base, divisor) {
  return {
    name,
    median_us: (whole.median_us - base.median_us) / divisor,
    p90_us: (whole.p90_us - base.p90_us) / divisor,
    iterations: whole.iterations,
  };
}

/** Prints the rows as a Markdown table. */
export function report(rows) {
  const value = (v) => (v === null ? 'n/a' : v.toFixed(2));
  console.log('| measurement | median µs | p90 µs |');
  console.log('|---|---:|---:|');
  for (const row of rows) console.log(`| ${row.name} | ${value(row.median_us)} | ${value(row.p90_us)} |`);
}

/** Writes the rows and their context to `<dir>/<file>` as JSON. */
export function writeResults(dir, file, payload) {
  mkdirSync(dir, { recursive: true });
  writeFileSync(join(dir, file), `${JSON.stringify({ node: process.version, arch: process.arch, ...payload }, null, 2)}\n`);
}

File: spikes/03-ephemeris-port/port/src/bench.rs
//! One timing helper for every Rust measurement of the spike: the median
//! of the best of `rounds` rounds, in microseconds, with the 90th
//! percentile, the same statistic the JavaScript and Dart harnesses of
//! spike 2 report.

use std::fmt::Write;
use std::time::Instant;

use serde::Serialize;

/// One measurement.
#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct Row {
    /// What was measured.
    pub name: String,
    /// The median of the best round, microseconds.
    pub median_us: f64,
    /// The 90th percentile of the best round, microseconds.
    pub p90_us: f64,
    /// Timed calls per round.
    pub iterations: usize,
    /// Rounds.
    pub rounds: usize,
}

/// Times `f` over `rounds` rounds of `iterations` calls after `warmup`
/// calls each, keeping the round with the lowest median.
pub fn bench<F: FnMut()>(
    name: &str,
    iterations: usize,
    warmup: usize,
    rounds: usize,
    mut f: F,
) -> Row {
    let mut best: Option<Row> = None;
    for _ in 0..rounds.max(1) {
        for _ in 0..warmup {
            f();
        }
        let mut times: Vec<f64> = (0..iterations.max(1))
            .map(|_| {
                let start = Instant::now();
                f();
                start.elapsed().as_secs_f64() * 1e6
            })
            .collect();
        times.sort_by(f64::total_cmp);
        let at = |q: f64| {
            #[allow(
                clippy::cast_possible_truncation,
                clippy::cast_sign_loss,
                clippy::cast_precision_loss,
                reason = "a quantile index"
            )]
            let i = ((q * times.len() as f64) as usize).min(times.len().saturating_sub(1));
            times.get(i).copied().unwrap_or_default()
        };
        let row = Row {
            name: name.to_string(),
            median_us: at(0.5),
            p90_us: at(0.9),
            iterations: iterations.max(1),
            rounds: rounds.max(1),
        };
        if best.as_ref().is_none_or(|b| row.median_us < b.median_us) {
            best = Some(row);
        }
    }
    best.unwrap_or(Row {
        name: name.to_string(),
        median_us: 0.0,
        p90_us: 0.0,
        iterations,
        rounds,
    })
}

/// The rows as a Markdown table.
#[must_use]
pub fn markdown(rows: &[Row]) -> String {
    let mut out = String::from("| measurement | median µs | p90 µs |\n|---|---:|---:|\n");
    for row in rows {
        let _ = writeln!(
            out,
            "| {} | {:.2} | {:.2} |",
            row.name, row.median_us, row.p90_us
        );
    }
    out
}
